### PR TITLE
remove DirectStake flow, fix pendingExitStake inflation, add invariant checker

### PR DESCRIPTION
**Pull request by hbdgr**
### **User description**
This PR introduces several improvements and fixes, identified during the HackenProof audit contest:

* Removes the deprecated **DirectStake** flow from contracts, tests, and documentation, simplifying the codebase and making the staking flow clearer
* Introduces the **InvariantChecker** and integrates it into the test suite
* Fixes edge cases causing `pendingExitStake` inflation
* Separates user pending exits from fee pending exits

___

### **PR Type**
Enhancement, Tests, Bug fix, Documentation

___

### **Description**
- Remove deprecated DirectStake flow

- Fix pendingExitStake inflation edge cases

- Separate fee vs user pending exits

- Add InvariantChecker and PoC tests

___

### Diagram Walkthrough

```mermaid
flowchart LR
  DirectStake["DirectStake flow (legacy)"] -- "removed" --> DepositFacet["DepositFacet"]
  DepositFacet -- "simplified events/APIs" --> IDeposit["IDeposit interface"]
  AllocationFacet["AllocationFacet"] -- "cap under loss + fee separation" --> StakeInfo["StakeInfo: +pendingExitFee"]
  HyperFactoryFacet["HyperFactoryFacet"] -- "no direct flag" --> VaultInfo["VaultInfo"]
  InvariantChecker["InvariantChecker (tests)"] -- "asserts stake/principal invariants" --> Tests["Unit & Foundry tests"]
  LockboxFacet["LockboxFacet"] -- "docs/comment cleanup" --> Messaging["Hyperlane routes"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td><strong>POC_H03.t.sol</strong><dd><code>Foundry PoC for underflow and failed redeems</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-717154cd6e012014ae63cd6f28dfe831c4d6c23958fe87412629d0179bae5433">+484/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>RedeemBurnsBeforeFee.t.sol</strong><dd><code>PoC: burn-before-fee leads to undeliverable message</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-dbcfc759f4cdb1ae8e0109769ee077bffb681c9a5efeff2afed95b4834ccf403">+282/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>InvariantChecker.sol</strong><dd><code>New invariant checker for stake/principal supply</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-5a2a6d5095df1d017f0d2817f7deacb801169d7a6d264b1ad4b6db8ceef4d196">+116/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>real-assets-redeem-harness.sol</strong><dd><code>Harness to assert redeem allowance and behavior</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-8bf7b81a60225bb442a0253b2077dbac2bff0b2519ffb464e2350dac9c39ab0e">+90/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Aera.test.ts</strong><dd><code>Refactor to base deploy, add invariants and fee-exit tests</code></dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-c911fbb2d38aba59611e16bfab174cb2e4f440730c65726302b67d7f5827f379">+390/-115</a></td>

</tr>

<tr>
  <td><strong>Strategy.test.ts</strong><dd><code>Refactor deploy, invariants, and rename strategy upgrade</code>&nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-f3691333309d201454d051bfe35499ce0f8e17d0598dad6d64bf85c30d08da47">+66/-50</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>Lockbox.test.ts</strong><dd><code>Refactor deploy; use lumia diamond routes; adjust events</code>&nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-6bf32ccf722dd4a074769d2444603d178707576c42fc36af1eb36d5566deefba">+56/-37</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>Staking.test.ts</strong><dd><code>Remove direct flow usage; adjust events and checks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-77fa49ac9fd5c2c1d692693b94f8b11fe7b00e73a1775514873d50050298fa2b">+50/-40</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>AllocationFacet.sol</strong><dd><code>Remove direct flow, fix exit cap, split fee exits</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-693e2d3ffbcf7103e2077db3c3cbcb6d2d62e7a0c7bd39dd6dfb20256d439c27">+27/-39</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>DepositFacet.sol</strong><dd><code>Remove direct deposit and typed events; adjust claims</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-b621111944793f39f0c5c5dcc43e628c368443739e27d9a07ef0cd05e9fae996">+10/-60</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>HyperFactoryFacet.sol</strong><dd><code>Remove direct flag; initialize StakeInfo with fee field</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-d54ef18bef7a611612eba919ce07aa15f630f5d36ec676e4a961401d89974468">+10/-26</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>IDeposit.sol</strong><dd><code>Drop DirectStake API and enum; event shape update</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-4cbb0ab7dced7cc74fc042e23add3486b987f73b49d6b6a4c6b72d43a5a899e0">+3/-33</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>LibHyperStaking.sol</strong><dd><code>Remove DirectStakeInfo; add pendingExitFee to StakeInfo</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-e91bc40a6a489969aee5a5d8902bcb5879b958e4874de2e0cc632800e0f7518a">+3/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>LumiaDiamond.ts</strong><dd><code>Ensure role read occurs after diamond cut</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-708632f5f0194814f4d69f0ad0c14fbe3f6ca3634cea9377e48f4486e2f702a8">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>IAllocation.sol</strong><dd><code>Remove direct-only errors; clarify docs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-543578c5673f443fda247f2a133951caea812d6d52594da25888de9abe61f592">+3/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>LockboxFacet.sol</strong><dd><code>Update comments to reflect non-direct handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-38de5ed27d598def25a61ecb13c74521696c3cf2b4ab3e680dcf9ecd4fb4c750">+1/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Formatting</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>SwapSuperStrategy.sol</strong><dd><code>Lint: disable var-name mixedcase</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-5259bc36d8fd68d69695641327d9adf8095b2b722c687b050ba2696e1c6f1a3b">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>hardhat.config.ts</strong><dd><code>Add global test setup and typings for forking config</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-57cf5378f8cab20b02b279097ba6fad08eb53f747d81770045dda9c5f95effe3">+4/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>14 files</summary><table>
<tr>
  <td><strong>IHyperFactory.sol</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-082000f9808d06d735fb2dc85f770755a2fd2311e3f32568bd162058bd9479a0">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>IStrategy.sol</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-d42c68a75af450df081750e8e29c55383aebebe932b5b8d458b49fcd0eb70e50">+0/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>AbstractStrategy.sol</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-ca5b75da66042f3c789f34c81495f302257b7f3776c19c1e97b2e417bf2fa11a">+0/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>DirectStakeStrategy.sol</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-ffe4a02f2ac95b613e1fbf44de0ece447bc7e712dead18cf4613f7d81e070d40">+0/-117</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>spec.md</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-4d2f1fceeeffb0bd025e300792b59495bb7215e19fdb143bebb083ada53730bd">+8/-22</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>DirectStakeStrategy.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-560c47a58fcb630b095f526edbe1fc925d422858dce608b59c2c9b7d016395c3">+0/-24</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>InvariantChecker.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-809edec00b9672fc8a41b268fb7423378b51d08c0af4d8b993b8b0db44fbd725">+15/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>real-assets-redeem2.sol.bak</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-435f1435de6cedb5ab2fa1f2ac526d3e8183317c962be82a2fc6e087d8d4161d">+115/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>setup.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-d28a461c46dd3d6440988800fbf174825ddbec8b60488095df612d33f1cad4df">+160/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>shared.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-f0abd6b40d03656b70fef247b9b3e96f5ae705ebfa15fbaf5d472dd366c1b993">+3/-132</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>CurveStrategy.test.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-e9749831b546fbbfe5fbcca5ab2eb50604e27f570701698fadcc8c0ff2a1ef2d">+44/-27</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>DirectStake.test.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-9b6ede7bf5be921b33f948bc4d09050d94412687c4f0e3044816ec75cec145b7">+0/-182</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>Superform.test.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-a72c16af620bdf4e906c2375484069dd047740fb711a44bd9ee7e34f63a50261">+42/-23</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>VaultShares.test.ts</strong></td>
  <td><a href="https://github.com/LumiaChain/HyperStaking-contracts/pull/5/files#diff-0f35510ce144af05ecf03bb38394b577fb303645e7da6588863ee9536dbf29e5">+47/-34</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

> [!NOTE]
> Removes the DirectStake flow, fixes exit accounting (split fee vs user exits, cap exits under loss), and adds an invariant checker with broad test/doc updates.
> 
> - **Core (contracts)**:
>   - Remove DirectStake logic across vault/strategy lifecycle; delete `DirectStakeStrategy` and related storage (`DirectStakeInfo`).
>   - Exit accounting hardening in `AllocationFacet._leave(...)`: cap allocation under loss, correct off‑by‑one from `previewAllocation`, and track fee exits separately via `StakeInfo{ pendingExitStake, pendingExitFee }`.
>   - Simplify deposit/withdraw paths in `DepositFacet`: drop direct deposit; adjust claim logic to decrement `pendingExitFee` for fee claims; emit streamlined events.
>   - `HyperFactoryFacet`: drop `direct` flag, initialize `StakeInfo` with `pendingExitFee`.
>   - `LockboxFacet`: comment/flow aligned to non‑direct; failed redeem handling unchanged; add re‑execute path.
> - **Interfaces & Events**:
>   - `IDeposit`: remove `DepositType` enum and direct API; update `StakeDeposit`/`WithdrawClaimed` event shapes.
>   - `IAllocation`: clarify docs; no direct‑only errors.
>   - `IStrategy`: remove `isDirectStakeStrategy`; keep `isIntegratedStakeStrategy`.
>   - `LibHyperStaking`: `VaultInfo` without `direct`; `StakeInfo` gains `pendingExitFee`.
> - **Testing & Tooling**:
>   - Add `contracts/test/InvariantChecker.sol` and ignition module; integrate into tests to assert stake vs principal supply invariants and pending‑exit bounds.
>   - Refactor tests to base deployment (`test/setup.ts`); update unit tests (Aera, Superform, Curve, Staking, VaultShares) to new APIs/events and fee‑exit flow; remove DirectStake tests.
>   - `hardhat.config.ts`: add global test setup; minor typing for forking.
>   - `ignition/modules/LumiaDiamond.ts`: ensure role read after diamond cut.
> - **Docs**:
>   - Update spec to remove direct staking references; clarify stake synchronization, redemption flow, and strategy flags.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2112858fc9c79883ebfe00dd00e6f882a77065a1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: contracts/test/InvariantChecker.sol [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            uint256 failedRedeemsStake = sumFailedRedeems(strategy);
+
+            // cannot have more pending exits than total stake
+            StakeInfo memory si = allocationFacet.stakeInfo(strategy);
+            require(
+                si.totalStake >= (si.pendingExitStake + failedRedeemsStake),
+                "inv: totalStake < pendingExitStake"
+            );
+
+            uint256 available = si.totalStake - (si.pendingExitStake + failedRedeemsStake);
```
[the remark this answers is not in the document]
Comment (hbdgr) [reply contracts/test/InvariantChecker.sol]: Actually, you're not right. If you look at the report function in `AllocationFacet`, the `feeAmount` is not added to `totalStake` so  available stake does not depend on it.

File: contracts/test/InvariantChecker.sol [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        for (uint256 i = 0; i < strategies.length; i++) {
+            address strategy = strategies[i];
+
+            uint256 failedRedeemsStake = sumFailedRedeems(strategy);
+
+            // cannot have more pending exits than total stake
+            StakeInfo memory si = allocationFacet.stakeInfo(strategy);
+            require(
+                si.totalStake >= (si.pendingExitStake + failedRedeemsStake),
+                "inv: totalStake < pendingExitStake"
```
[the remark this answers is not in the document]
Comment (hbdgr) [reply contracts/test/InvariantChecker.sol]: As previous, wrong, because `totalStake` does not include fee.
